# Advanced Python
(by Joe Marini in LinkedIn Learning)

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Python-Language-Features" data-toc-modified-id="Python-Language-Features-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Python Language Features</a></span><ul class="toc-item"><li><span><a href="#Python-Coding-Style" data-toc-modified-id="Python-Coding-Style-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Python Coding Style</a></span></li><li><span><a href="#Truth-Value-Testing" data-toc-modified-id="Truth-Value-Testing-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Truth Value Testing</a></span></li><li><span><a href="#String-vs.-Bytes" data-toc-modified-id="String-vs.-Bytes-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>String vs. Bytes</a></span></li><li><span><a href="#Template-Strings" data-toc-modified-id="Template-Strings-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Template Strings</a></span></li></ul></li><li><span><a href="#Built-in-Functions" data-toc-modified-id="Built-in-Functions-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Built-in Functions</a></span><ul class="toc-item"><li><span><a href="#Utilities" data-toc-modified-id="Utilities-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Utilities</a></span></li><li><span><a href="#Transformers" data-toc-modified-id="Transformers-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Transformers</a></span><ul class="toc-item"><li><span><a href="#filter" data-toc-modified-id="filter-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span><code>filter</code></a></span></li><li><span><a href="#map" data-toc-modified-id="map-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span><code>map</code></a></span></li></ul></li><li><span><a href="#Itertools" data-toc-modified-id="Itertools-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Itertools</a></span><ul class="toc-item"><li><span><a href="#itertools.cycle" data-toc-modified-id="itertools.cycle-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span><code>itertools.cycle</code></a></span></li><li><span><a href="#itertools.count" data-toc-modified-id="itertools.count-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span><code>itertools.count</code></a></span></li><li><span><a href="#itertools.accumulate" data-toc-modified-id="itertools.accumulate-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span><code>itertools.accumulate</code></a></span></li><li><span><a href="#itertools.chain" data-toc-modified-id="itertools.chain-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span><code>itertools.chain</code></a></span></li><li><span><a href="#dropwhile-and-takewhile" data-toc-modified-id="dropwhile-and-takewhile-2.3.5"><span class="toc-item-num">2.3.5&nbsp;&nbsp;</span><code>dropwhile</code> and <code>takewhile</code></a></span></li></ul></li></ul></li><li><span><a href="#Advanced-Python-Function" data-toc-modified-id="Advanced-Python-Function-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Advanced Python Function</a></span><ul class="toc-item"><li><span><a href="#Function-Documentation-Strings" data-toc-modified-id="Function-Documentation-Strings-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Function Documentation Strings</a></span></li><li><span><a href="#Variable-Argument-Lists" data-toc-modified-id="Variable-Argument-Lists-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Variable Argument Lists</a></span></li><li><span><a href="#Lambda-Functions" data-toc-modified-id="Lambda-Functions-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Lambda Functions</a></span></li><li><span><a href="#Keyword-only-Arguments" data-toc-modified-id="Keyword-only-Arguments-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Keyword-only Arguments</a></span></li></ul></li><li><span><a href="#Advanced-Collections" data-toc-modified-id="Advanced-Collections-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Advanced Collections</a></span><ul class="toc-item"><li><span><a href="#Advanced-Collections-Overview" data-toc-modified-id="Advanced-Collections-Overview-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Advanced Collections Overview</a></span></li><li><span><a href="#namedtuple" data-toc-modified-id="namedtuple-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span><code>namedtuple</code></a></span></li><li><span><a href="#defaultdict" data-toc-modified-id="defaultdict-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span><code>defaultdict</code></a></span></li><li><span><a href="#Counter" data-toc-modified-id="Counter-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span><code>Counter</code></a></span></li><li><span><a href="#OrderedDict" data-toc-modified-id="OrderedDict-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span><code>OrderedDict</code></a></span></li><li><span><a href="#deque" data-toc-modified-id="deque-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span><code>deque</code></a></span></li></ul></li><li><span><a href="#Advanced-Classes-and-Objects" data-toc-modified-id="Advanced-Classes-and-Objects-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Advanced Classes and Objects</a></span><ul class="toc-item"><li><span><a href="#Advanced-Classes-Overview" data-toc-modified-id="Advanced-Classes-Overview-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Advanced Classes Overview</a></span></li><li><span><a href="#Defining-Enumerations" data-toc-modified-id="Defining-Enumerations-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Defining Enumerations</a></span></li><li><span><a href="#Class-String-Values" data-toc-modified-id="Class-String-Values-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Class String Values</a></span></li><li><span><a href="#Computed-Attributes" data-toc-modified-id="Computed-Attributes-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Computed Attributes</a></span></li><li><span><a href="#Object-Operations" data-toc-modified-id="Object-Operations-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Object Operations</a></span></li><li><span><a href="#Object-Comparisons" data-toc-modified-id="Object-Comparisons-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>Object Comparisons</a></span></li></ul></li><li><span><a href="#Using-Logging" data-toc-modified-id="Using-Logging-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Using Logging</a></span><ul class="toc-item"><li><span><a href="#Overview-of-Logging" data-toc-modified-id="Overview-of-Logging-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Overview of Logging</a></span></li><li><span><a href="#Basic-Logging" data-toc-modified-id="Basic-Logging-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Basic Logging</a></span></li><li><span><a href="#Custom-Logging" data-toc-modified-id="Custom-Logging-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Custom Logging</a></span></li></ul></li><li><span><a href="#Python-Comprehensions" data-toc-modified-id="Python-Comprehensions-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Python Comprehensions</a></span><ul class="toc-item"><li><span><a href="#What-Are-Comprehensions" data-toc-modified-id="What-Are-Comprehensions-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>What Are Comprehensions</a></span></li><li><span><a href="#List-Comprehensions" data-toc-modified-id="List-Comprehensions-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>List Comprehensions</a></span></li><li><span><a href="#Dictionary-Comprehensions" data-toc-modified-id="Dictionary-Comprehensions-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Dictionary Comprehensions</a></span></li><li><span><a href="#Set-Comprehensions" data-toc-modified-id="Set-Comprehensions-7.4"><span class="toc-item-num">7.4&nbsp;&nbsp;</span>Set Comprehensions</a></span></li></ul></li><li><span><a href="#The-End" data-toc-modified-id="The-End-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>The End</a></span></li></ul></div>

## Python Language Features

### Python Coding Style

__Important:__
- Code gets read more often than it gets written.
- Adhering to Pep8 makes it easier to read/write code for you and others. 
    
__Rules:__
- Indent code with spaces instead of tabs (use 4 spaces for each indentation level). 
- Limit lines to 79 characters (including indents)
- Limit comments/doc to 72. 
- Functions and Classes should be separated by 2 blank lines.
- Within classes, 1 blank line should separate methods.
- No spaces around function calls, indexes, arguments.
    
__White Space Conventions:__

Wrong | Correct
--- | --- 
`spam( ham[ 1 ], {eggs: 2 }` | `spam(ham[1], {eggs: 2})`
`fn (arg)` | `fn(arg)`
`dict [key] = list [index]` | `dict[key] = list[index]`
`x   = 1` | `x = 1`
`z = x * x + y * y` | `z = x*x + y*y`
`i=i+1` | `i = i + 1`

### Truth Value Testing

__Important:__
- Any object in python can be tested for boolean value. By default all objects are `True` unless if `len()=0` or `bool()=False`.
    
__Evaluates To False:__
1. `False` or `None`
2. Numeric zero values: `0, 0.0, or 0j`
3. `Decimal(0)` or `Franction(0, x)`
4. Empty sequences: `(), [], {}`
5. Empty sets and ranges: `set(), range(0)`
6. Custom objects with over written `bool() or len()`

In [1]:
# Default Objects: 
print('bool test: ', bool(object))

bool test:  True


In [2]:
# Custom Objects (with 0 len method): 
class MyClass:
    def __len__(self):
        return 0
    
myClass = MyClass()
print('bool test: ', bool(myClass))

bool test:  False


In [3]:
# Custom Objects (with 1 len method but False bool): 
class MyClass:
    def __len__(self):
        return 1
    def __bool__(self):
        return False

myClass = MyClass()
print('bool test: ', bool(myClass)) # Priority to bool method!

bool test:  False


__Basic Boolean Operations:__ 

In [4]:
print("True and True:", True and True)
print("True and False:", True and False)

True and True: True
True and False: False


In [5]:
print("True or True:", True or True)
print("True or False:", True or False)

True or True: True
True or False: True


In [6]:
print("not True:", not True)
print("not False:", not False)

not True: False
not False: True


### String vs. Bytes

- Encode: __str  ---> byte__
- Decode: __byte ---> str__

In [7]:
b = bytes([0x41, 0x42, 0x43, 0x44])
b

b'ABCD'

In [8]:
s = 'This is a String'
s

'This is a String'

In [9]:
# b + s # TypeError: can't concat str to bytes
b_s = b.decode()
s + b_s

'This is a StringABCD'

In [10]:
s_b = s.encode()
b + s_b

b'ABCDThis is a String'

In [11]:
s.encode('utf-32')

b'\xff\xfe\x00\x00T\x00\x00\x00h\x00\x00\x00i\x00\x00\x00s\x00\x00\x00 \x00\x00\x00i\x00\x00\x00s\x00\x00\x00 \x00\x00\x00a\x00\x00\x00 \x00\x00\x00S\x00\x00\x00t\x00\x00\x00r\x00\x00\x00i\x00\x00\x00n\x00\x00\x00g\x00\x00\x00'

### Template Strings

__Important:__
- There are at least 4 ways to format strings in Python.
- `Template` from `string` module is more secure. 

In [12]:
"You are a {0} {1}".format("Bat", "Man")

'You are a Bat Man'

In [13]:
from string import Template
temp = Template("You are a ${Attr} ${Gen}")
temp.substitute(Attr="Bat", Gen="Man")

'You are a Bat Man'

In [14]:
data = {'Attr': 'Bat', 'Gen': 'Man'}
temp.substitute(data)

'You are a Bat Man'

## Built-in Functions

### Utilities

- Can be accessed in Python documentation under [built-in-functions](https://docs.python.org/3/library/functions.html#built-in-functions).

In [15]:
ls = [1, 2, 3, 0, 5, 6]
print('any:', any(ls)) # True if at least one element is not zero. 
print('all:', all(ls)) # True only if all elements are not zero. 
print('min:', min(ls))
print('max:', max(ls))
print('sum:', sum(ls))

any: True
all: False
min: 0
max: 6
sum: 17


### Transformers

#### `filter`

__Important:__
It creates an iterator that filters out values from a given __sequence__ based on a given __function__.

In [16]:
def odd(x):
    if x%2 == 0:
        return False
    return True

list(filter(odd, range(10)))

[1, 3, 5, 7, 9]

In [17]:
def lower(x):
    if x.isupper():
        return False
    return True

list(filter(lower, "abcDEFGHijk"))

['a', 'b', 'c', 'i', 'j', 'k']

#### `map`

__Important:__
It creates an iterator that takes one __sequence__ or more and convert each item based on a given __function__.

In [18]:
def square(x):
    return x**2

list(map(square, range(10)))

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

In [19]:
def toGrade(x):
    if (x >= 90):
        return 'A'
    elif (x >= 80 and x < 90):
        return 'B'
    elif (x >= 70 and x < 80):
        return 'C'
    elif (x >= 65 and x < 70):
        return 'D'
    else:
        return 'F'
    
list(map(toGrade, [81, 68, 94, 78, 61]))

['B', 'D', 'A', 'C', 'F']

### Itertools

In [20]:
import itertools

#### `itertools.cycle`

In [21]:
myCycle = itertools.cycle(range(4))
print("cycle 1:", next(myCycle))
print("cycle 2:", next(myCycle))
print("cycle 3:", next(myCycle))
print("cycle 4:", next(myCycle))
print("cycle 5:", next(myCycle))
print("cycle 6:", next(myCycle))

cycle 1: 0
cycle 2: 1
cycle 3: 2
cycle 4: 3
cycle 5: 0
cycle 6: 1


#### `itertools.count`

In [22]:
myCount = itertools.count(10, 10)
print("Count 1:", next(myCount))
print("Count 2:", next(myCount))
print("Count 3:", next(myCount))
print("Count 4:", next(myCount))
print("Count 5:", next(myCount))
print("Count 6:", next(myCount))

Count 1: 10
Count 2: 20
Count 3: 30
Count 4: 40
Count 5: 50
Count 6: 60


#### `itertools.accumulate`

In [23]:
list(itertools.accumulate([1, 3, 2, 5, 0]))

[1, 4, 6, 11, 11]

In [24]:
list(itertools.accumulate([1, 3, 2, 5, 0], max))

[1, 3, 3, 5, 5]

In [25]:
list(itertools.accumulate([1, 3, 2, 5, 0], min))

[1, 1, 1, 1, 0]

#### `itertools.chain`

In [26]:
list(itertools.chain("ABCD", "1234"))

['A', 'B', 'C', 'D', '1', '2', '3', '4']

#### `dropwhile` and `takewhile`

In [27]:
def cut(x):
    return x < 40

In [28]:
list(itertools.dropwhile(cut, [10, 20, 30, 40, 50, 40, 30, 20, 10]))

[40, 50, 40, 30, 20, 10]

- Note: start the cut once the condition is met. 

In [29]:
list(itertools.takewhile(cut, [10, 20, 30, 40, 50, 40, 30, 20, 10]))

[10, 20, 30]

- Note: stop the cut once the condition is met. 

## Advanced Python Function

### Function Documentation Strings

__Rules:__
- First line should be a summary sentence. 
- Modules: list important classes and functions. 
- Classes: list important methods. 
- Functions: 
    - list important parameters and explain them (each one in a separate line).
    - explain the output if the function returns a value. 
    - explain exceptions if the function raises any. 

(for more info check [Pep257](https://www.python.org/dev/peps/pep-0257/))

In [30]:
def myFunc(arg1, arg2=None):
    """ myFunction(arg1, arg2=None) -> list
    Paramaters:
        arg1: the first argument. 
        arg2: the second argument. Defaults to None. 
    """
    return []
    
print(myFunc.__doc__)

 myFunction(arg1, arg2=None) -> list
    Paramaters:
        arg1: the first argument. 
        arg2: the second argument. Defaults to None. 
    


### Variable Argument Lists

__Important:__ 
- This allows to define a function that can take a variable number of arguments.
- This approach is useful when the number of arguments is relatively small.

In [31]:
def addition(*args):
    result = 0
    for arg in args:
        result += arg
    return result

In [32]:
addition(10, 10, 20, 20)

60

In [33]:
addition(*[10, 10, 20, 20]) # star (*) is used to consider each item as a seperate argument. 

60

### Lambda Functions

__Important:__
- This approach is known in other language as "anonymous function".
- Definition: `lambda x: func(x)`.

In [34]:
def CelsiusToFahrenheit(temp):
    return (temp * 9/5) + 32

In [35]:
ctemps = [0, 12, 34, 100]
list(map(CelsiusToFahrenheit, ctemps)) 

[32.0, 53.6, 93.2, 212.0]

__Note: since the function is simple, it is preferable to use in-line `lambda`.__

In [36]:
list(map(lambda t: (t*9/5)+32, ctemps)) # better for readability.

[32.0, 53.6, 93.2, 212.0]

### Keyword-only Arguments

__Important:__
- This approach allows some arguments (after *) to be used only with their keyword.
- This approach is used to insure quality in the code. 

In [37]:
def myFunc(arg1, arg2, *, Flage=False):
    pass

In [38]:
myFunc(1, 2)
# myFunc(1, 2, True) # TypeError: myFunc() takes 2 positional arguments but 3 were given
myFunc(1, 2, Flage=True)

## Advanced Collections

In [39]:
import collections

### Advanced Collections Overview

__Basic Collections:__
- `list`: mutable sequence of values. 
- `tuple`: immutable sequence of values. 
- `set`: mutable unordered sequence of distinct values. 
- `dict`: mutable collection of key-value pairs. 

__Advanced Collections:__
- `namedtuple`: tuple with named fields. 
- `defaultdict`: dictionaries with special properties.
- `Counter`: counts distinct values.
- `OrderedDict`: dictionaries with special properties.
- `deque`: double ended list object. 

### `namedtuple`

__Important:__
- `namedtuple` helps to increase the code readability.
- In this case, we can't use default argument values. If the data has a large number of optional/default values, then it is better to use the regular tuple. 

In [40]:
p = (10, 20)
print('p', p)
print('p[0]:', p[0])
print('p[1]:', p[1])

p (10, 20)
p[0]: 10
p[1]: 20


In [41]:
Point = collections.namedtuple("Point", ['x', 'y'])
p1 = Point(10, 20)
p2 = Point(30, 40)
print(p1, p2)
print('p1[0]', p1[0])
print('p1.x:', p1.x)

Point(x=10, y=20) Point(x=30, y=40)
p1[0] 10
p1.x: 10


In [42]:
p1 = p1._replace(x=100)
print(p1)

Point(x=100, y=20)


### `defaultdict`

In [43]:
fruits = ['apple', 'pear', 'orange', 'banana', 'apple', 'banana', 'grape', 'banana']
fruitCounter = {}
for fruit in fruits:
    # fruitCounter[fruit]+=1 # KeyError: 'apple' does not exit!
    if fruit in fruitCounter.keys():
        fruitCounter[fruit] += 1
    else: 
        fruitCounter[fruit] = 1
        
fruitCounter

{'apple': 2, 'pear': 1, 'orange': 1, 'banana': 3, 'grape': 1}

__But this code is unnecessarily long!__

In [44]:
fruitCounter = collections.defaultdict(int) # defaults to 0 for each new key
for fruit in fruits: 
    fruitCounter[fruit] += 1
fruitCounter

defaultdict(int, {'apple': 2, 'pear': 1, 'orange': 1, 'banana': 3, 'grape': 1})

In [45]:
fruitCounter = collections.defaultdict(lambda: 10) # defaults to 10 for each new key
for fruit in fruits: 
    fruitCounter[fruit] += 1
fruitCounter

defaultdict(<function __main__.<lambda>()>,
            {'apple': 12, 'pear': 11, 'orange': 11, 'banana': 13, 'grape': 11})

### `Counter`

__Important:__
- `Counter` is a `dict` sub-class for counting hashable objects. 
- `Counter` is similar to `defaultdict`but with additional features to work with numbers and items. 

In [46]:
class1 = ['Bob', 'Frank', 'James', 'James']
class2 = ['Bill', 'Frank', 'James', 'Joe']
c1 = collections.Counter(class1)
c2 = collections.Counter(class2)

print('c1:', c1)
print('c2:', c2)
print(c1 & c2) # count for items that exit in both.

c1: Counter({'James': 2, 'Bob': 1, 'Frank': 1})
c2: Counter({'Bill': 1, 'Frank': 1, 'James': 1, 'Joe': 1})
Counter({'Frank': 1, 'James': 1})


In [47]:
print("c1['James']:", c1['James'])
print("c1 # of students: ", len(c1))

c1['James']: 2
c1 # of students:  3


In [48]:
print(c1.most_common(3))
print(c2.most_common(3))

[('James', 2), ('Bob', 1), ('Frank', 1)]
[('Bill', 1), ('Frank', 1), ('James', 1)]


In [49]:
c1.update(class2)
c1

Counter({'Bob': 1, 'Frank': 2, 'James': 3, 'Bill': 1, 'Joe': 1})

In [50]:
c1.subtract(class2)
c1

Counter({'Bob': 1, 'Frank': 1, 'James': 2, 'Bill': 0, 'Joe': 0})

### `OrderedDict`

__Important:__
- `OrderedDict` is a dict that tracks the order in which items are inserted/sorted. 

In [51]:
ordict1 = collections.OrderedDict({'a': 1, 'b': 2, 'c': 3})
ordict2 = collections.OrderedDict({'a': 1, 'c': 3, 'b': 2})
dict1 = dict(ordict1)
dict2 = dict(ordict2)

In [52]:
print(ordict1 == ordict2) # order is important
print(dict1 == dict2) # order is not important
print(dict1 == ordict2) # order is not important

False
True
True


In [53]:
teams = {('Barca', (22, 8)),
         ('Juvi', (20, 11)),
         ('Liver', (23, 5))}

teams_sorted = sorted(teams, key=lambda t: sum(t[1]), reverse=True) # sort by total sum or points
#teams_sorted = sorted(teams, key=lambda t: t[1][0], reverse=True) # sort by first amount of points
teams_sorted

[('Juvi', (20, 11)), ('Barca', (22, 8)), ('Liver', (23, 5))]

In [54]:
for name, points in teams:
    print(name, sum(points), end=' | ')

Juvi 31 | Barca 30 | Liver 28 | 

In [55]:
teams_ordict = collections.OrderedDict(teams_sorted) # order is maintaned!
teams_ordict

OrderedDict([('Juvi', (20, 11)), ('Barca', (22, 8)), ('Liver', (23, 5))])

### `deque`

__Important:__
- `deque` stands for double-ended queue and pronounced as "deck".
- `deque` allows to append or pop data from either side.
- `deque` is designed to be memory efficient. 
- `deque` can be initialized as an empty iterable object.
- `deque` cab be specified to max length. 
- `deque` contains:
    - `.append()` and `.appendleft()`
    - `.pop()` and `.popleft()`
    - `.rotate()` or `.rotate(1)`: rotate to right (default) by one step
    - `.rotate(-1)`: rotate to left
    - `.rotate(n)`: rotate to right n steps
    - `.rotate(-n)`: rotate to left n steps

In [56]:
import string

letters = string.ascii_uppercase
letters

'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [57]:
letters_deque = collections.deque(letters)
print(letters_deque)

deque(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'])


In [58]:
letters_deque.pop()
letters_deque.popleft()
print(letters_deque)

deque(['B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y'])


In [59]:
letters_deque.append(1)
letters_deque.appendleft(0)
print(letters_deque)

deque([0, 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 1])


In [60]:
letters_deque.rotate(2)
print(letters_deque)

deque(['Y', 1, 0, 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X'])


In [61]:
import time
rotations = range(10)
for i in rotations:
    print('rotation {}:'.format(i), letters_deque, end='\r')
    letters_deque.rotate()
    time.sleep(0.4)

## Advanced Classes and Objects

### Advanced Classes Overview

__Important:__
- Python is very customizable! we can:
    - create enumerations.
    - customize string and byte representation of objects (`str`, `repr`),
    - define computed and default attributes.
    - control how objects logically compared to each other (`>, <, ==, >=, <=`).
    - give objects numeric-like behavior (`+, -`). 

### Defining Enumerations

__Important:__
- `Enum` is used to assign easy to read names as a hash values to increase code readability and iterate over. 
- `Enum` is a class that can be used to create object which have unique keys.

In [62]:
from enum import Enum, unique, auto

In [63]:
class Fruit(Enum):
    APPLE = 1
    BANANA = 2
    ORANGE = 3
    TOMATO = 4
    RED = 1
    # APPLE = 8 # TypeError: Attempted to reuse key: 'APPLE' 

print(Fruit.APPLE)
print(type(Fruit.APPLE))
print(repr(Fruit.APPLE))
print(Fruit.APPLE.name, Fruit.APPLE.value)

Fruit.APPLE
<enum 'Fruit'>
<Fruit.APPLE: 1>
APPLE 1


In [64]:
@unique # a class decorator to prevent duplicate values
class Fruit(Enum):
    APPLE = 1
    BANANA = 2
    ORANGE = 3
    TOMATO = 4
    # RED = 1 # ValueError: duplicate values found in <enum 'Fruit'>: RED -> APPLE 
    # APPLE = 8 # TypeError: Attempted to reuse key: 'APPLE' 

print(Fruit.APPLE)

Fruit.APPLE


In [65]:
@unique # a class decorator to prevent duplicate values
class Fruit(Enum):
    APPLE = auto() # to assign numbers according to order. 
    BANANA = auto()
    ORANGE = auto()
    TOMATO = auto()
    # RED = 1 # ValueError: duplicate values found in <enum 'Fruit'>: RED -> APPLE 
    # APPLE = 8 # TypeError: Attempted to reuse key: 'APPLE' 

print(Fruit.APPLE.value)
print(Fruit.BANANA.value)
print(Fruit.ORANGE.value)
print(Fruit.TOMATO.value)

1
2
3
4


In [66]:
# Using enum as a hash value: 
myFruits = {}
myFruits[Fruit.BANANA] = "Yes!"
print(myFruits[Fruit.BANANA])

Yes!


### Class String Values

__Important:__
- Converting different kind of values into a string representation is a very common operation.
- Python provides a custom class with full control on how the string form will be presented. 
    - `__str__(self)`: used with `str(object)`, `print(object)`, `"{}".format(object)`
    - `__repr__(self)`: used with `repr(onject)` 
        - used for debugging to recreate the same object with the same value.
        - if we overwrite `__repr__`and not `__str__`, `__repr__`will also overwrite `__str__`.
    - `__format__(self)`: used with `format(object, format_spec)`
    - `__bytes__(self)`: used with `bytes(object)`

In [67]:
class Person():
    def __init__(self, fname, lname, age):
        self.fname = fname
        self.lname = lname
        self.age = age
    
    def __repr__(self):
        return "<Person Class - fname: {}, lname: {}, age: {}>".format(self.fname, self.lname, self.age)
    
    def __str__(self):
        return "Person ({} {} is {} years old)".format(self.fname, self.lname, self.age)
    
    def __bytes__(self):
        return "Person: {}, {}, {}".format(self.fname, self.lname, self.age).encode('utf-8')

In [68]:
person1 = Person("Zaka", "Grin", 31)
print(person1)
print(str(person1))
print(repr(person1))
print(bytes(person1))

Person (Zaka Grin is 31 years old)
Person (Zaka Grin is 31 years old)
<Person Class - fname: Zaka, lname: Grin, age: 31>
b'Person: Zaka, Grin, 31'


### Computed Attributes

__Important:__
- used to retrieve and set attribute values.
- `__getattribute__(self, attr)`: to get attributes values `object.attr` (called unconditionally each time).
- `__getattr__(self, attr)`: to get attributes values with `object.attr` (when `__getattribute__()` is not defined).
- `__setattr__(self, attr, val)`: to set attributes with `object.attr = val`.
- `__delattr__(self, attr)`: to del attribures with `del(object.attr)`.
- `__dir__(self)`: to return string with available attributes and methods with `dir(object)`.

In [69]:
class Color():
    def __init__(self, red, green, blue):
        self.red = red
        self.green = green
        self.blue = blue
    
    def __getattr__(self, attr):
        if attr == 'rgb':
            return (self.red, self.green, self.blue)
        elif attr == 'bgr':
            return (self.blue, self.green, self.red)
        elif attr == 'hex':
            return "#{0:02x}{1:02x}{2:02x}".format(self.red, self.green, self.blue)
        else:
            raise AttributeError # for unknown attr, araise an Error!
            
    def __setattr__(self, attr, val):
        if attr == 'rgb':
            self.red = val[0]
            self.green = val[1]
            self.blue = val[2]
        elif attr == 'bgr':
            self.blue = val[0]
            self.green = val[1]
            self.red = val[2]
        else:
            super().__setattr__(attr, val) # otherwise, __getattr__() will be called which cause an Error! 
        
    def __dir__(self):
        return ('red', 'green', 'blue', 'rgb', 'bgr', 'hex')

In [70]:
color = Color(red=10, green=20, blue=40)
print(color.rgb)
print(color.bgr)
print(color.hex)

(10, 20, 40)
(40, 20, 10)
#0a1428


In [71]:
color.bgr = (20, 40, 50)
print(color.bgr)
print(color.rgb)
print(color.hex)

(20, 40, 50)
(50, 40, 20)
#322814


In [72]:
dir(color)

['bgr', 'blue', 'green', 'hex', 'red', 'rgb']

### Object Operations

__Important:__
- `__add__(self, other)`: used with addition (`self + other`).
- `__sub__(self, other)`: used with subtraction (`self - other`).
- `__mul__(self, other)`: used with multiplication (`self * other`).
- `__div__(self, other)`: used with division (`self / other`).
- `__Floordiv__(self, other)`: used with integer division (`self // other`).
- `__pow__(self, other)`: used with power (`self ** other`).
- `__and__(self, other)`: used with `and` operator (`self and other` or `self & other`).
- `__or__(self, other)`: used with `or` operator (`self or other` or `slef | other`).

(for inplace operations, `i` is added before the operation name)
- `__iadd__(self, other)`: used with inplace addition (`self += other`).
- `__isub__(self, other)`: used with inplace subtraction (`self -= other`).

...
- `__iand__(self, other)`: used with inplace `and` operator (`self &= other`).
- `__ior__(self, other)`: used with inplace `or` operator (`self |= other`).

(See [Python Documentation Data Model](https://docs.python.org/3/reference/datamodel.html)). 

In [73]:
class Point():
    def __init__(self, x, y):
        self.x = x
        self.y = y
    
    def __repr__(self):
        return "<Point x:{}, y:{}>".format(self.x, self.y)
    
    def __add__(self, other):
        return Point(self.x+other.x, self.y+other.y)
    
    def __sub__(self, other):
        return Point(self.x-other.x, self.y-other.y)
    
    def __iadd__(self, other):
        self.x += other.x
        self.y += other.y
        return self

In [74]:
p1 = Point(10, 20)
p2 = Point(30, 40)
print(p1+p2)
print(p1-p2)
p1 += p2
print(p1)

<Point x:40, y:60>
<Point x:-20, y:-20>
<Point x:40, y:60>


### Object Comparisons

__Important:__
- `__gt__(self, other)`: used with `>` operator (`self > other`).
- `__ge__(self, other)`: used with `>=` operator (`self >= other`).
- `__lt__(self, other)`: used with `<` operator (`self < other`).
- `__le__(self, other)`: used with `<=` operator (`self <= other`).
- `__eq__(self, other)`: used with `==` operator (`self == other`).
- `__ne__(self, other)`: used with `!=` operator (`self != other`).

In [75]:
class Employee():
    def __init__(self, name, level, years):
        self.name = name
        self.level = level
        self.years = years
        
    def __gt__(self, other):
        if self.level == other.level:
            return self.years > other.years
        return self.level > other.level
    
    def __ge__(self, other):
        if self.level == other.level:
            return self.years >= other.years
        return self.level >= other.level
    
    def __lt__(self, other):
        if self.level == other.level:
            return self.years < other.years
        return self.level < other.level

    def __le__(self, other):
        if self.level == other.level:
            return self.years <= other.years
        return self.level <= other.level
    
    def __eq__(self, other):
        return self.level == other.level
    
    def __repr__(self): # to pass a string value
        return "Employee: name={}, level={}, years={}".format(self.name, self.level, self.years)
    
    #def __str__(self): # to print (if not defined then __repr__()). 
    #    return "Employee: name={}, level={}, years={}".format(self.name, self.level, self.years)

In [76]:
omar = Employee('omer', 3, 3)
zaka = Employee('Zaka', 3, 4)
print('> ', omar > zaka)
print('>=', omar >= zaka)
print('< ', omar < zaka)
print('<=', omar <= zaka)
print('==', omar == zaka)

>  False
>= False
<  True
<= True
== True


In [77]:
dept = []
dept.append(Employee('Sim', 5, 9))
dept.append(Employee('Dos', 4, 12))
dept.append(Employee('Jan', 6, 6))
dept.append(Employee('Rob', 7, 11))
dept.append(Employee('Dur', 6, 12))

In [78]:
dept[0] > dept[2]

False

In [79]:
dept[4] < dept[3]

True

In [80]:
dept

[Employee: name=Sim, level=5, years=9,
 Employee: name=Dos, level=4, years=12,
 Employee: name=Jan, level=6, years=6,
 Employee: name=Rob, level=7, years=11,
 Employee: name=Dur, level=6, years=12]

In [81]:
sorted(dept)

[Employee: name=Dos, level=4, years=12,
 Employee: name=Sim, level=5, years=9,
 Employee: name=Jan, level=6, years=6,
 Employee: name=Dur, level=6, years=12,
 Employee: name=Rob, level=7, years=11]

## Using Logging

In [82]:
import logging

def read_log_file(path):
    file = open(path, mode='r')
    for i, line in enumerate(file):
        print(i, '\t', line.strip())
    file.close()

### Overview of Logging

__Important:__
- Logging is an important tool in the development process where events are recorded for later analysis. 
- Logging events helps to diagnose problems. 
- If we use remote/external server, it is difficult to debug without recording these events in files which can be done by logging. 
- Some errors happens only occasionally during the program life time. These error can be captured by logging.
- Logging makes it easy to understand events happened while the program executed which allows us to track errors. 
- Logging has a highly customizable output. Therefore, everyone should use them (but in fact they do not).

### Basic Logging

__Basic Levels:__
- `logging.debug("message")`: used for debugging
- `logging.info("message")`: used to record general info.
- `logging.warning("message")`: used to indicate an approaching problem.
- `logging.problem("message")`: used to record a problem.
- `logging.critical("message")`: used to record a serious error.

__Config:__
- `logging.basicConfig(level=logging.DEBUG, filename='output.log', filemode='w')`: 
    - `level=logging.DEBUG`: to set the minimum required level to `debug` level (default is `warning`).
    - `filename`: file name including the path.
    - `filemode`: 
        - `a`: append to the file.
        - `w`: write on top of an existing file (i.e. previous info will be deleted).
- Note that for levels we use uppercase (e.g `DEBUG, INFO`) while in functions we use lowercase (e.g. `debug, info`).

In [83]:
read_log_file('output1.log')

0 	 INFO:root:info message
1 	 WARNING:root:warning message
2 	 ERROR:root:error message
3 	 CRITICAL:root:critical message
4 	 INFO:root:info message


### Custom Logging

__Important:__
- `logging.basicConfig()` can also take:
    - `format=formatstr`: to format messages.
        - `formatstr=`
            - `%(asctime)s`: human readable log date.
            - `%(filename)s`: lof filename.
            - `%(funcName)s`: log function name.
            - `%(levelname)s`: log level name.
            - `%(levelno)d`: log level number.
            - `%(lineno)d`: log line number.
            - `%(message)s`: log message.
            - `%(module)s`: log module name.
    - `datefmt=datestr`: to format date string.
        - `datestr="%m/%d/%y %I:%M:%S %p"`

In [84]:
extData = {'user':'zg@go.com'}
formatstr = '%(asctime)s (%(funcName)s - Line: %(lineno)d - User: %(user)s) - %(levelname)s: %(message)s'
datestr = '%m/%d/%y %I:%M:%S %p'

logging.basicConfig(level=logging.DEBUG,
                    filename='output2.log',
                    filemode='w',
                    format=formatstr,
                    datefmt=datestr)

logging.debug("debug message", extra=extData)
logging.info("info message", extra=extData)
logging.warning("warning message", extra=extData)

In [85]:
read_log_file('output2.log')

0 	 02/07/21 11:53:43 AM (<module> - Line: 11 - User: zg@go.com) - DEBUG: debug message
1 	 02/07/21 11:53:43 AM (<module> - Line: 12 - User: zg@go.com) - INFO: info message
2 	 02/07/21 11:53:43 AM (<module> - Line: 13 - User: zg@go.com) - WARNING: warning message


## Python Comprehensions

### What Are Comprehensions

__Important:__
- Comprehensions can be applied to: `list, dict, set`.
    - `list`comprehensions can do exactly like `map(lambda)` but with simpler code. 
        - `list(map(FahrenheitToCelsius, [32, 65, 104, 212]))`.
        - `list(map(lambda t: (t*9/5)+32, [32, 65, 104, 212]))`.
        - `[(t*9/5)+32 for t in [32, 65, 104, 212]]`.

### List Comprehensions

In [86]:
numbers = range(1, 21)
evens = list(filter(lambda num: num%2==0, numbers))
odds = list(filter(lambda num: num%2!=0, numbers))

In [87]:
evenSequered = list(map(lambda num: num**2, evens))
evenSequered

[4, 16, 36, 64, 100, 144, 196, 256, 324, 400]

In [88]:
# filter to select only: [6, 8, 10, 12, 14]
evenSequered = list(map(lambda num: num**2, filter(lambda num: num>4 and num<16, evens)))
evenSequered

[36, 64, 100, 144, 196]

In [89]:
# list comprehensions is much simpler than map: 
evenSequered = [num**2 for num in evens if num>4 and num<16]
evenSequered

[36, 64, 100, 144, 196]

### Dictionary Comprehensions

In [90]:
ctemp = [0, 12, 34, 100]
tempDict = {t: (t*9/5)+32 for t in ctemp if t<100}
print(tempDict)
print(tempDict[12])

{0: 32.0, 12: 53.6, 34: 93.2}
53.6


In [91]:
team1 = {'Jan': 24, 'Felix': 18, 'Jim': 58}
team2 = {'Zak': 40, 'Marco': 50, 'Erik': 49}

newteam = {k:v for team in (team1, team2) for k,v in team.items()}
newteam

{'Jan': 24, 'Felix': 18, 'Jim': 58, 'Zak': 40, 'Marco': 50, 'Erik': 49}

### Set Comprehensions

In [92]:
ctemps = [5, 10, 12, 14, 10, 23, 41]
ftemps1 = [(t*9/5)+32 for t in ctemps] # list: order and duplicates
ftemps2 = {(t*9/5)+32 for t in ctemps} # set: no order no duplicates
print(ftemps1)
print(ftemps2)

[41.0, 50.0, 53.6, 57.2, 50.0, 73.4, 105.8]
{73.4, 41.0, 105.8, 50.0, 53.6, 57.2}


In [93]:
text = "Bla Bla"
chars = {c.upper() for c in text if not c.isspace()}
chars

{'A', 'B', 'L'}

## The End

__Important:__
- Look at the following courses:
    - Learning Python GUI Programming.
    - Python: Data Analysis
- Keep your skills up to date by reading the python documentation [https://www.python.org](https://www.python.org/).

In [94]:
print('The End!')

The End!
